src/magic_folder/magic_file.py

```diff
@@ -1286,15 +1389,28 @@ def _done_working(self):
 
     _conflicted.upon(
         _conflict_resolution,
-        enter=_uploading,
-        outputs=[_begin_upload],
+        enter=_creating_snapshot,
+        outputs=[_working, _status_upload_queued, _create_local_snapshot],
         collector=_last_one,
     )
     _conflicted.upon(
         _remote_update,
-        enter=_conflicted,
-        outputs=[],  # probably want to .. do something? remember it?
+        enter=_downloading,
+        outputs=[_working, _status_download_queued, _begin_download]
     )
+    # XXX this is trix-y .. the user could be changing the local
+    # contents because they'll ultimately be doing a "resolve" (with
+    # those contents) __OR__ they could be naively simply editing the
+    # file more.
+    #
+    # ...so I think we _do_ want to queue a local-update: an incoming
+    # resolution may happen, and if we have no local change: great,
+    # apply it. But if we _do_ have a local change, we'd want to
```

Comment:
This sounds about right to me.

In such instances, though, I wonder if would it be worth perhaps uniquely identifying each individual "instance" of conflict -- perhaps associating each instance with a UUID or something -- so that we know or can track from which instance of conflict was ultimately resolved (and "conflict again" only if the other instance(s) were resolved)?

I was thinking about this originally in the context of https://github.com/tahoe-lafs/magic-folder/issues/755, and wondering if, for example, it might make sense to have a "conflict-_updated_" event (in order to signal, say, that another change has happened on top of an already-conflicted file that did _not_ resolve the conflict, or maybe to signal that a two-participant conflict has become a three-participant one -- e.g., going from `"participants": ["Alice", "Bob"]` to `"participants": ["Alice", "Bob", "Charlie"]`) versus, instead, just emitting another "conflict-_detected_" message (and considering all conflicts associated with a given `relpath` as being "resolved" if/when one of them is).

(I don't have particularly strong opinions here as I haven't thought very far past the simpler "two participants, one conflict" case...)

Reply:
we do already track multiple "instances" in the sense that we track which other participants are currently conflicted (which can be 0..n)

the only case i can think of might be where e.g. "alice" is conflicted, and then updates her copy further (so on "our" side we'd still be conflicted).

Is there anything that should behave differently in this case? (Clients _could_ I think already figure out where the conflict "first" happened if they wanted, by examining the version tree).

Reply:
>the only case i can think of might be where e.g. "alice" is conflicted, and then updates her copy further (so on "our" side we'd still be conflicted).

It also still seems possible (I believe?) to have participants resolve a conflict in diverging ways that would require additional intervention. For example, Alice might accept/prefer Bob's copy, while Charlie ("at the same time") accepts/prefers Alice's original copy (i.e., prior to her accepting of Bob's).

I suppose the questions here are more UX-oriented ones -- about whether and how users need to be exposed (and/or make decisions about) each "intermediate" instance in a multi-conflict scenario. Presumably, whichever snapshots are "the latest" probably matter most here (especially since past snapshots are preserved in the event they are needed).

Anyway, I don't think there's anything that needs to be done about this right now; sorry for the digression!

Reply:
Yes, two resolutions "at the same time" are possible -- they should immediately produce another conflict (probably need _lots_ more tests like this :thinking: ). However, this isn't logically "another instance" of an existing conflict, but a new conflict (between the Snapshot that Alice produced and the Snapshot that Charlie produced as resolutions).